src/main/java/graphql/annotations/connection/ConnectionDataFetcher.java

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright 2016 Yurii Rashkovskii
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+package graphql.annotations.connection;
+
+import graphql.relay.Connection;
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+import java.util.Optional;
+
+import static graphql.annotations.ReflectionKit.constructNewInstance;
+
+public class ConnectionDataFetcher<T> implements DataFetcher<Connection<T>> {
+    private final Class<? extends ConnectionFetcher<T>> connection;
+    private final DataFetcher<PaginatedData<T>> actualDataFetcher;
```

Comment:
PaginatedData is tied to EnhancedConnectionFetcher, but here we should not assume we have DataFetcher<PaginatedData<T>> , only a DataFetcher<?> that would be understood by the ConnectionFetcher implementation. (also see comment below)